viewing_party/party.py

```diff
@@ -1,23 +1,226 @@
+                user_data["watchlist"].pop(movie_index)
+    return user_data
 
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
+def get_watched_avg_rating(user_data):
+    '''
+    Input: user_data is a dictionary with key "watched" and values of a list of movie dictionaries
+    (containing title, gener, and rating), representing movies the user has watched. 
+    Output: function finds the average rating of all movies in watched list and returns that average rating. 
+    '''
+    sum = 0
+    for movie in user_data["watched"]:
+        sum += movie["rating"]
+    try:
+        average = sum / len(user_data["watched"])
+    except ZeroDivisionError:
+        average = 0.0
+    return average
+
+def get_most_watched_genre(user_data):
+    '''
+    Input: user_data is a dictionary with a "watched" list of movie dictionaries. Each movie has a key "genre"
+    Output: Function determines which genre is most frequent and returns that genre value. If "watched"
+    is an empty list, function returns None.
+    '''
+    genre_list = []
+    genre_count = {}
+
+    if user_data["watched"] == []:
+        return None
+
+    for movie in range(len(user_data["watched"])):
+        genre_list.append(user_data["watched"][movie]['genre'])
+        
+    for i in genre_list:
+        if i not in genre_count:
+            count = 1
+            genre_count[i] = count
+        else:
+            genre_count[i] = count + 1
+    
+    max_genre = max(genre_count, key =genre_count.get)
+    return max_genre
 
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
+def get_unique_watched(user_data):
+    ''' 
+        Input: user_data is a dictionary with "watched" (list of movie dictionaries) and "friends"
+        (with a list of movie dictionaries of what friends have watched)
+        Output: returns list of dictionaries of unique movies the user has watched
+    '''
+    user_set = user_watched_list(user_data)
+    friends_set = friend_watched_list(user_data)
+    unique_movies_set = user_set - friends_set
+    unique_movies_list = list(unique_movies_set)
+
+    result_unique_movies = []
+
+    for unique_movie in unique_movies_list:
+        for movie in user_data["watched"]:
+            if unique_movie == movie["title"]:
+                result_unique_movies.append(movie)
+    
+    return result_unique_movies
+    
+def user_watched_list(user_data):
+    '''
+    Helper function
+    Input: user_data a dictionary with "watched" and "friends" keys contianing lists of movies
+    Output: returns set of movies the user has watched, but friends have not
+    '''
+    user_watch_list = []
+    
+    for movie in user_data["watched"]:
+        for key,value in movie.items():
+            if key == "title":
+                user_watch_list.append(value)

```

Comment:
Remember to use dictionary keys to access values. This loop can be refactored to be:
```
for movie in user_data["watched"]:
    user_watch_list.append(movie["title"])
```